Introduce `IntegrationPattern` abstraction

artembilan opened the pull request:
* Add `IntegrationPattern` contract to implement on the target components
which represent particular EIP
* Add `IntegrationPatternType` with an internal `IntegrationPatternCategory`
to return from the component implementing `IntegrationPattern`
* Parse `IntegrationPatternType` in the `IntegrationNode` for potential
use on the UI for drawing a particular icon

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
Introduce `IntegrationPattern` abstraction

* Add `IntegrationPattern` contract to implement on the target components
which represent particular EIP
* Add `IntegrationPatternType` with an internal `IntegrationPatternCategory`
to return from the component implementing `IntegrationPattern`
* Parse `IntegrationPatternType` in the `IntegrationNode` for potential
use on the UI for drawing a particular icon

artembilan commented:
/CC @BoykoAlex

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/IntegrationPattern.java:
```diff
@@ -0,0 +1,39 @@
+package org.springframework.integration;
+
+/**
+ * A marker interface indicating that component implements some
```
a component

garyrussell replied:
Not strictly a "marker" - they are usually empty.

How about "Indicates that a component ..." ?

garyrussell reviewed (changes requested):
Maybe I am missing something but I don't see enough information about how the UI will choose an icon.

artembilan commented:
see an `IntegrationNode` changes. There is already an `integrationPatternType` property.
Why this cannot be mapped to icon on the UI?
I fully don't understand the reason to mix UI attributes in our code.
That is really not our model responsibility to dictate the icon.
An `integrationPatternType`  alway can simply be mapped to some icon on UI as one-to-one relationship.

What am I missing?

Thanks

BoykoAlex commented:
Great news! Adding @kdvolder as well.

garyrussell commented:
It's not clear what the category is for; if it's not exposed on the graph.

garyrussell commented:
My confusion is AMH exposes itself as service activator whereas it can be that, an outbound gateway, etc, etc

artembilan committed:
* More pattern representations

artembilan committed:
* Clean up Checkstyle
* Fix JavaDocs
* Add `integrationPatternCategory` assertion into the `IntegrationGraphServerTests`

artembilan committed:
* Add more IntegrationPattern implementations
* Provide some delegation and overriding logic whenever we have
components wrapping

artembilan commented:
Added more pattern implementations.

TODO:

1. Inbound Gateways
2. Condition Outbound gateway/channel adapter

artembilan committed:
* Fix unused imports

artembilan committed:
* Add `inbound_gateway` pattern indicator
* Add conditional on `expectReply` to indicate a component as an
`IntegrationPatternType.outbound_channel_adapter` or
`IntegrationPatternType.outbound_gateway`
* Make some code clean up in affected classes

artembilan commented:
From my point of view it is ready for merge for UI project.
We may consider to add more IEP indicators in other components, like `MessageStore`.
We don't have a place to add a hook for `Idempotent Receiver` (although just an `IdempotentReceiverInterceptor` could have that) and for `Routing Slip` which is just a header in the message.
All of those could be done on demand.

Anyway before merging I would like to discuss why should we treat a `@MessagingGateway` differently from  any other `Inbound Gateway` implementation. It just does exactly the same: wraps an inbound data into a `Message` and sends it into a channel.
So, from the UI perspective it really looks the same. Not sure why should we extract it into a separate integration pattern and also introduce a new UI icon for it...

Thanks

artembilan commented:
However I see that there enough attention is paid for that pattern in the book: https://www.enterpriseintegrationpatterns.com/patterns/messaging/MessagingGateway.html ...

artembilan commented:
OK. I've changed my mind: will show a `@MessagingGateway` as a `gateway` pattern...

artembilan committed:
* Add a `gateway` type for `@MessagingGateway`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/IntegrationPatternType.java:
```diff
@@ -0,0 +1,161 @@
+				bridge),
+
+		message_routing(
+				splitter,
```
This is tricky - a splitter is not really routing in any way, but maybe it can't be considered a transformer either, although it does transform one message into many.

Maybe it needs its own category?

artembilan replied:
I placed it into a category where it belongs in EIP: https://www.enterpriseintegrationpatterns.com/patterns/messaging/Sequencer.html

> Messaging Patterns » Message Routing » Splitter

garyrussell replied:
OK

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/IntegrationPatternType.java:
```diff
@@ -0,0 +1,161 @@
+		message_routing(
+				splitter,
+				filter,
+				aggregator,
```
ditto

maybe splitter. aggregator, scatter_gather, resequencer in a new category?

artembilan replied:
DITTO

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractReplyProducingMessageHandler.java:
```diff
@@ -91,6 +92,11 @@ protected ClassLoader getBeanClassLoader() {
 
+	@Override
+	public IntegrationPatternType getIntegrationPatternType() {
+		return IntegrationPatternType.outbound_gateway;
```
This is not intuitive; I would expect the default to be service_activator; I am guessing you did this to avoid having to patch every outbound gateway, which is ok, but perhaps a comment here to explain why this is the default might be useful.

artembilan replied:
Right. I did an opposite: whenever it is not an `outbound gateway`, I did an explicit type: `service activator`, `transformer`, `splitter` etc.

garyrussell replied:
Right, but I think it warrants a comment (not java doc, just `/* ... */`.

garyrussell reviewed (commented):
Good work; just a few comments for your consideration - also perhaps `IntegrationGraphServerTests` needs to be enhanced to check the graph is populated properly?

artembilan commented:
> also perhaps `IntegrationGraphServerTests` needs to be enhanced to check the graph is populated properly?

What do you mean, Gary?

I did some changes over there to be sure that new attributes are populated.
Perhaps not all of them, of course, but still...

Thanks

garyrussell commented:
Sorry; I somehow skipped over the test.

artembilan commented:
Hm.. I've looked into a docs for this and it turns out that we definitely need bump graph version - `private static final float GRAPH_VERSION = 1.2f;`, because of these new attributes.
So, I'm finishing the task anyway, but again: it sounds more like a feature for next minor version - `5.3`.

Any other thoughts?

garyrussell commented:
No more thoughts; we should get this to the tools team sooner rather than later, though.

artembilan committed:
* Comment the reason for `outbound_gateway` type in the `AbstractReplyProducingMessageHandler`
* Bump `IntegrationGraphServer.GRAPH_VERSION`
* Add new attributes into graph sample in the `graph.adoc`
* Document an `IntegrationPattern`

artembilan committed:
* Apply changes for version 5.3

artembilan commented:
OK. I've this change to version `5.3`. 
Will update *What's New* when the respective branch is ready.

artembilan committed:
* Rebased into `5.3-WIP`
* Add a `whats-new.adoc` note about an `IntegrationPattern`

artembilan commented:
OK. So, this is rebased onto `5.3-WIP`. Ready for final review and merge.
Therefore this new feature is going to be available in the published `BUILD-SNAPSHOT` for version `5.3.0`.

Thanks

garyrussell commented on src/reference/asciidoc/whats-new.adoc:
```diff
@@ -15,6 +15,12 @@ If you are interested in more details, see the Issue Tracker tickets that were r
 === New Components
 
+[[x5.2-rsocket-support]]
+==== Integration Pattern
```
5.3 - will fix on merge after travis is green.

artembilan replied:
Doh... And not `rsocket-support` 😄 

Thanks

garyrussell replied:
Also not rsocket - fixed.

garyrussell commented on src/reference/asciidoc/graph.adoc:
```diff
@@ -251,6 +257,9 @@ The preceding gateway produces nodes similar to the following:
+The version 5.3 has introduced an `IntegrationPattern` abstraction and all out-of-the-box components, which belongs to particular EI pattern implements this abstraction and provides an `IntegrationPatternType` enum value.
+This information can be useful for some categorizing logic in the target application or being exposed into the graph node it can be used on UI to determine how to draw the provided component.
+
 === Integration Graph Controller
```
Merged as 71224bc1c6738c1f4eb877c0d3d4763c94b0eb02 with minor doc polishing and added new elements to the graph example above this paragraph.
